exercise1/submissions/john_furr.py

```diff
@@ -0,0 +1,119 @@
+
+    return mean_data
+
+def calculate_stats_data(mean_data, datacols=None):
+    # Input integrity checks
+    if not datacols:
+        raise ValueError("You must provide datacols")
+
+    #   Part 2:   
+    #   Calculate the highest, lowest, median, and quartile ratings for each of the 9 questions. Generate a CSV file stats.csv
+    stats_data = pd.DataFrame(
+        index=('count', 'mean', 'std', 'min', '25%', '50%', '75%', 'max'),
+        columns=datacols,
+    )
+       
+    # Technically this produces numbers that are very close to the output
+    # though there are slight rounding errors.  I'm guessing that in practice
+    # this would be the preferred method, but in an effort to exactly match
+    # the example output a more verbose method is taken. 
+    
+    #mean_data = mean_data.convert_objects() # convert object data to numeric data
+    #stats_data = mean_data.describe()
+       
+    # Verbose method to exaclty match output precision
+    for colname in datacols:
+        stats_data[colname]['count'] = "%.1f"%mean_data[colname].count()
+        stats_data[colname]['mean'] = mean_data[colname].mean()
+        stats_data[colname]['std'] = mean_data[colname].std()
+        stats_data[colname]['min'] = mean_data[colname].min()
+        stats_data[colname]['25%'] = mean_data[colname].quantile(0.25)
+        stats_data[colname]['50%'] = mean_data[colname].quantile(0.50)
+        stats_data[colname]['75%'] = mean_data[colname].quantile(0.75)
+        stats_data[colname]['max'] = mean_data[colname].max()
+
+    return stats_data
+
+if __name__ == '__main__':
+    #
+    # Part 1:  Calculate mean data for specfic fields at the Client Code level.
+    #
+    # NOTE Instructions said to use:  'undrorg'  but the output file example contains 'impsust'
+    # NOTE In order to match the output file I am using impsust, but changing this would be a trivial change
+    #      in the usecols tuple below.
+    usecols=('fdntext', 'fldimp', 'undrfld', 'advknow', 'pubpol', 'comimp', 'undrwr', 'undrsoc', 'orgimp', 'impsust')
+    datacols=usecols[1:]
+    mean_data = calculate_mean_data(
+        client_col='fdntext', 
+        infile=join(INDIR, "xl.csv"),
+        usecols=usecols,
+        datacols=datacols
+    )
+    # Write output file and ignore row names/indexes
+    outfile = join(OUTDIR, "mean_comp.csv")
+    mean_data.to_csv(outfile, index=False)
+
+    # Check that the output file is identical with example file
+    assert filecmp.cmp(outfile, join(OUTDIR, "mean.csv")) is True
+
+    #
+    # Part 2)  Calculate addition stats aginst mean data
+    #
+    stats_data = calculate_stats_data(mean_data=mean_data, datacols=datacols)
+
+    # write stats data, this time with row names
+    stat_output = join(OUTDIR, "stat_comp.csv")
+    stats_data.to_csv(stat_output, index=True)
+
+    # check that output file is identical with example file
+    assert filecmp.cmp(stat_output, join(OUTDIR, "stats.csv")) is True
```

Comment:
Testing is important in things that we do here. Typically we use `unittest` but you're right that in this case, setting up `unittest` is an overkill for this exercise.

Didn't know you can compare two files directly with `filecmp`, so that's a nice trick.